barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -264,4 +264,141 @@ namespace avm_alu(256);
+    pol A_SUB_B_HI = a_hi - b_hi - borrow;
+
+    // Check that (a <= b) by checking (b_lo >= a_lo && b_hi >= a_hi) || (b_lo > a_lo && b_hi >= a_hi)
+    // First condition is if borrow = 0, second condition is if borrow = 1;
+    pol B_SUB_A_LO = b_lo - a_lo + borrow * 2 ** 128;
+    pol B_SUB_A_HI = b_hi - a_hi - borrow;
+
+    pol commit res_lo;
+    pol commit res_hi;
+
+    // If this is a LT operation, we already swapped the inputs so the result of ic is still correct
+    // If this is a LTE operation, we invert the value of ic.
+    pol IS_GT = lt_sel * ic + (1 - ic) * lte_sel;
+
+    res_lo = (A_SUB_B_LO * IS_GT + B_SUB_A_LO * (1 - IS_GT)) * cmp_sel;
+    res_hi = (A_SUB_B_HI * IS_GT + B_SUB_A_HI * (1 - IS_GT)) * cmp_sel;
+
+
+    // ========= RANGE OPERATIONS ===============================
+    // Each call to LT/LTE requires 5x 256-bit range checks. We keep track of how many are left here.
+    pol commit rng_chk_remaining;
+
+    // TODO: combine into 1 equation, left separate for debugging
+    // the number of range checks must decrement by 1 until it is equal to 0;
+    (rng_chk_remaining' - rng_chk_remaining + 1) * rng_chk_remaining = 0;
+    // if this row is a comparison operation, the next range_check_remaining value is set to 4
+    // it is not set to 5 since we do 1 as part of the comparison.
+    (rng_chk_remaining' - 4) * cmp_sel = 0;
+    
+    pol commit rng_chk_sel; // A boolean representing if we have range checks remaining
+    rng_chk_sel * (1 - rng_chk_sel);
+    rng_chk_sel * cmp_sel = 0; // If we have remaining range checks, we cannot have cmp_sel set
+
+    // Needed for equality checking // reuse selector op_eq_diff_inv
+    // pol commit rng_chk_inv; 
+
+    // rng_chk_sel = 1 when rng_chk_remaining != 0 and rng_chk_sel = 0 when rng_chk_remaining = 0;
+    rng_chk_remaining * ((1 - rng_chk_sel) * (1 -  op_eq_diff_inv) +  op_eq_diff_inv) - rng_chk_sel = 0;
+
+    // We perform a range check if we have some range checks remaining or we are performing a comparison op
+    pol RNG_CHK_OP = (rng_chk_sel + cmp_sel);
+
+    // Perform 128-bit range check on lo part
+    a_lo = (u8_r0 + u8_r1 * 2**8 + u16_r0 * 2**16 + 
+           u16_r1 * 2**32 + u16_r2 * 2**48 + 
+           u16_r3 * 2**64 + u16_r4 * 2**80 + 
+           u16_r5 * 2**96 + u16_r6 * 2**112) * (RNG_CHK_OP);
+
+    // Perform 128-bit range check on hi part
+    a_hi = (u16_r7 + u16_r8 * 2**16 + 
+           u16_r9 * 2**32 + u16_r10 * 2**48 + 
+           u16_r11 * 2**64 + u16_r12 * 2**80 + 
+           u16_r13 * 2**96 + u16_r14 * 2**112) * (RNG_CHK_OP);
+
+    // Shift all elements "across" by 2 columns
+    a_lo' = b_lo * rng_chk_sel';
+    a_hi' = b_hi * rng_chk_sel';
+    b_lo' = p_sub_a_lo * rng_chk_sel';
+    b_hi' = p_sub_a_hi * rng_chk_sel';
+    p_sub_a_lo' = p_sub_b_lo * rng_chk_sel';
+    p_sub_a_hi' = p_sub_b_hi * rng_chk_sel';
+    p_sub_b_lo' = res_lo * rng_chk_sel';
+    p_sub_b_hi' = res_hi * rng_chk_sel';
```

Comment:
We still need to perform 8-bit/16-bit range checks for the registers u8_XXX and u16_XXX.
I am fine to introduce them once this PR is merged as part of the CAST opcode work.
Please write a TODO in this file.